xmldumps-backup/mwbzutils/docs/dumpbz2filefromoffset.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.4.
.TH DUMPBZ2FILEFROMOFFSET "1" "July 2013" "dumpbz2filefromoffset 0.0.4" "User Commands"
.SH NAME
dumpbz2filefromoffset \- Write MediaWiki XML pages from bzip2 file starting from offset
.SH SYNOPSIS
.B dumpbz2filefromoffset
[\fI--version|--help\fR]
.br
.B dumpbz2filefromoffset
\fI<infile> <offset>\fR
.SH DESCRIPTION
Find the first bz2 block in a file after the specified offset, uncompress
and write contents from that point on to stdout, starting with the first
<page> tag encountered.
.PP
The starting <mediawiki> tag and the <siteinfo> header from the file will
be written out first.
.PP
Note that some bytes from the very last block may be lost if the blocks are
not byte\-aligned. This is due to the bzip2 crc at the eof being wrong.
.PP
Exits with BZ_OK on success, various BZ_ errors otherwise.
.SH OPTIONS

Flags:
.TP
\fB\-h\fR, \fB\-\-help\fR
Show this help message
.TP
\fB\-v\fR, \fB\-\-version\fR
Display the version of this program and exit
.PP
Arguments:
.TP
<infile>
Name of the file to check
.TP
<offset>
byte in the file from which to start processing
.SH AUTHOR
Written by Ariel T. Glenn.
.SH "REPORTING BUGS"
Report bugs in dumpbz2filefromoffset to <https://phabricator.wikimedia.org/>.
.PP
.br
See also checkforbz2footer(1), dumplastbz2block(1), findpageidinbz2xml(1),
recompressxml(1), writeuptopageid(1)
.SH COPYRIGHT
Copyright \(co 2011, 2012, 2013 Ariel T. Glenn.  All rights reserved.
.PP
This program is free software: you can redistribute it and/or modify it
under the  terms of the GNU General Public License as published by the
Free Software Foundation, either version 2 of the License, or (at your
option) any later version.
.PP
This  program  is  distributed  in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program.  If not, see <http://www.gnu.org/licenses/>


xmlfileutils/CHANGELOG
29 May 2013 Kent L. Miller <wpmirrordev@gmail.com>

   Release 0.0.2-1

    * mwxml2sql.c:
      Doc: show_version() rewritten for compatibility with help2man.
      Doc: usage() rewritten for compatibility with help2man.
      Doc: man page generated with help2man
    * sql2txt.c:
      Bug fixed: long options, --sqlfile and --textfile, did not work.
      Doc: show_version() written for compatibility with help2man.
      Doc: usage() rewritten for compatibility with help2man.
      Doc: man page generated with help2man
      Option: added -V, --version option for use with help2man
    * sqlfilter.c:
      Doc: show_version() written for compatibility with help2man.
      Doc: usage() rewritten for compatibility with help2man.
      Doc: man page generated with help2man
      Option: created -w, --version option for use by help2man
    * Makefile:
      Debian: build and deinstall targets written; clean target improved
      Debian: install target rewritten to use: install $(DESTDIR)/...
      Debian: dist renamed `mediawiki-mwxml2sql-0.0.2.tar.gz'
      Doc: man pages generated using `help2man'
      MWXML2SQLVERS for Makefile dist version and for VERSION in source
    * mwxml2sql.h:
      VERSION passed from Makefile

22 May 2013 Ariel T. Glenn <ariel@wikimedia.org>

    * mwxmlelts.c:
      handle text with <deleted /> instead of content
      workaround for XML files that may contain duplicate text ids

21 May 2013 Ariel T. Glenn <ariel@wikimedia.org>

    * mwxmlelts.c:
      convert revision timestamp from datestring format to db format

10 Apr 2013 Ariel T. Glenn <ariel@wikimedia.org>

    Release 0.0.2

    * sqlfilter: filters sql table dumps against e.g. list of pageids
      (and the column number of the pageid in the sql row) and writes
      only those tuples that match
    * wikiretriever.py: retrieves page titles and or content from a wiki
    * wikicontent2sql.py: downloads all pages containing a given template,
      swapping out talk pages for their corresponding pages (eg File talk
      for File etc.), converts them to page, revision and table sql,
      and filters downloaded sql table dumps according to the page ids
      of the downloaded content, producing a pile of sql files for import
    * package: various bugfixes

25 Jan 2013 Ariel T. Glenn <ariel@wikimedia.org>

   Release 0.0.1

   Features:
    *  Reads stub and text content xml files in parallel, producing separate
       sql files for inserts to page, revision and text tables
    *  Reads bz2, gz or plain text content, writes bz2, gz or plaintext output
       based on extensions of stub/page content files and output filename
    *  Reads xml dumps with schema 0.5 through 0.8, writes for MW 1.5 through
       1.20 (tested only for 0.7 in and MW 1.20 out)


xmldumps-backup/mwbzutils/dumpbz2filefromoffset.c
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <regex.h>
#include <getopt.h>
#include "mwbzutils.h"

void usage(char *message) {
  char * help =
"Usage: dumpbz2filefromoffset [--version|--help]\n"
"   or: dumpbz2filefromoffset <infile> <offset>\n\n"
"Find the first bz2 block in a file after the specified offset, uncompress\n"
"and write contents from that point on to stdout, starting with the first\n"
"<page> tag encountered.\n\n"
"The starting <mediawiki> tag and the <siteinfo> header from the file will\n"
"be written out first.\n\n"
"Note that some bytes from the very last block may be lost if the blocks are\n"
"not byte-aligned. This is due to the bzip2 crc at the eof being wrong.\n\n"
"Exits with BZ_OK on success, various BZ_ errors otherwise.\n\n"
"Options:\n\n"
"Flags:\n\n"
"  -h, --help       Show this help message\n"
"  -v, --version    Display the version of this program and exit\n\n"
"Arguments:\n\n"
"  <infile>         Name of the file to check\n"
"  <offset>         byte in the file from which to start processing\n\n"
"Report bugs in dumpbz2filefromoffset to <https://bugzilla.wikimedia.org/>.\n\n"
"See also checkforbz2footer(1), dumplastbz2block(1), findpageidinbz2xml(1),\n"
    "recompressxml(1), writeuptopageid(1)\n\n";
  if (message) {
    fprintf(stderr,"%s\n\n",message);
  }
  fprintf(stderr,"%s",help);
  exit(-1);
}

void show_version(char *version_string) {
  char * copyright =
"Copyright (C) 2011, 2012, 2013 Ariel T. Glenn.  All rights reserved.\n\n"
"This program is free software: you can redistribute it and/or modify it\n"
"under the  terms of the GNU General Public License as published by the\n"
"Free Software Foundation, either version 2 of the License, or (at your\n"
"option) any later version.\n\n"
"This  program  is  distributed  in the hope that it will be useful, but\n"
"WITHOUT ANY WARRANTY; without even the implied warranty of \n"
"MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General\n"
"Public License for more details.\n\n"
"You should have received a copy of the GNU General Public License along\n"
"with this program.  If not, see <http://www.gnu.org/licenses/>\n\n"
    "Written by Ariel T. Glenn.\n";
  fprintf(stderr,"dumpbz2filefromoffset %s\n", version_string);
  fprintf(stderr,"%s",copyright);
  exit(-1);
}

/* 
   dump the <mediawiki> header (up through
   </siteinfo> close tag) found at the 
   beginning of xml dump files. 
   returns:
      0 on success,
      -1 on error
*/
int dump_mw_header(int fin) {
  int res;
  regmatch_t *match_siteinfo;
  regex_t compiled_siteinfo;
  int length=5000; /* output buffer size */
  char *siteinfo = "  </siteinfo>\n";

  buf_info_t *b;
  bz_info_t bfile;

  int firstpage = 1;
  int done = 0;
  bfile.initialized = 0;

  res = regcomp(&compiled_siteinfo, siteinfo, REG_EXTENDED);

  match_siteinfo = (regmatch_t *)malloc(sizeof(regmatch_t)*1);

  b = init_buffer(length);
  bfile.bytes_read = 0;
  bfile.position = (off_t)0;

  while ((get_buffer_of_uncompressed_data(b, fin, &bfile, FORWARD)>=0) && (! bfile.eof) && (!done)) {
    /* fixme either we don't check the return code right or we don't notice no bytes read or we don't clear the bytes read */
    if (bfile.bytes_read) {
      if (firstpage) {
	if (bfile.bytes_read >= 11 && !memcmp((char *)b->next_to_read,"<mediawiki ",11)) {
	  /* good, write it and loop and not firstpage any more */
	  if (b->bytes_avail) {
	    if (regexec(&compiled_siteinfo, (char *)b->next_to_read,  2,  match_siteinfo, 0 ) == 0) {
	      fwrite(b->next_to_read,match_siteinfo[0].rm_eo, 1, stdout);
	      b->next_to_read = b->end;
	      b->bytes_avail = 0;
	      b->next_to_fill = b->buffer; /* empty */
	      bfile.strm.next_out = (char *)b->next_to_fill;
	      bfile.strm.avail_out = b->end - b->next_to_fill;
	      done++;
	    }
	    else {
	      fwrite(b->next_to_read,b->bytes_avail,1,stdout);
	      b->next_to_read = b->end;
	      b->bytes_avail = 0;
	      b->next_to_fill = b->buffer; /* empty */
	      bfile.strm.next_out = (char *)b->next_to_fill;
	      bfile.strm.avail_out = b->end - b->next_to_fill;
	    }
	  }  
	}
	else {
	  fprintf(stderr,"missing mediawiki header from bz2 xml file\n");
	  return(-1);
	}
	firstpage = 0;
      }
      else { /* not firstpage */
	if (regexec(&compiled_siteinfo, (char *)b->next_to_read,  2,  match_siteinfo, 0 ) == 0) {
	  fwrite(b->next_to_read,match_siteinfo[0].rm_eo, 1, stdout);
	  b->next_to_read = b->end;
	  b->bytes_avail = 0;
	  b->next_to_fill = b->buffer; /* empty */
	  bfile.strm.next_out = (char *)b->next_to_fill;
	  bfile.strm.avail_out = b->end - b->next_to_fill;
	  done++;
	}
	else {
	  /* could have the first part of the siteinfo tag... so copy up enough bytes to cover that case */
	  if (b->bytes_avail> 12) {
	    /* write everything that didn't match, but leave 12 bytes, to stdout */
	    fwrite(b->next_to_read,b->bytes_avail - 12,1,stdout);
	    move_bytes_to_buffer_start(b, b->next_to_read + b->bytes_avail - 12, 12);
	    bfile.strm.next_out = (char *)b->next_to_fill;
	    bfile.strm.avail_out = b->end - b->next_to_fill;
	  }
	  else {
	    if (buffer_is_empty(b)) {
	      bfile.strm.next_out = (char *)b->buffer;
	      bfile.strm.avail_out = bfile.bufout_size;
	      b->next_to_fill = b->buffer; /* empty */
	    }
	    else {
	      /* there were only 12 or less bytes so just save em don't write em to stdout */
	      move_bytes_to_buffer_start(b, b->next_to_read, b->bytes_avail);
	      bfile.strm.next_out = (char *)b->next_to_fill;
	      bfile.strm.avail_out = b->end - b->next_to_fill;
	    }
	  }
	}
      } /* end notfirstpage */
    }
  }
  if (!done) {
    fprintf(stderr,"incomplete or no mediawiki header found\n");
    return(-1);
  }
  else {
    return(0);
  }
}

/* 
   find the first page id after position in file 
   decompress and dump to stdout from that point on
   returns:
      0 on success,
      -1 on error
*/
int dump_from_first_page_id_after_offset(int fin, off_t position) {
  int res;
  regmatch_t *match_page;
  regex_t compiled_page;
  int length=5000; /* output buffer size */
  char *page = "  <page>";

  buf_info_t *b;
  bz_info_t bfile;

  int firstpage = 1;

  bfile.initialized = 0;

  res = regcomp(&compiled_page, page, REG_EXTENDED);

  match_page = (regmatch_t *)malloc(sizeof(regmatch_t)*1);

  b = init_buffer(length);
  bfile.bytes_read = 0;
  bfile.position = position;

  while ((get_buffer_of_uncompressed_data(b, fin, &bfile, FORWARD)>=0) && (! bfile.eof)) {
    /* fixme either we don't check the return code right or we don't notice no bytes read or we don't clear the bytes read */
    if (bfile.bytes_read) {
      if (firstpage) {
	if (regexec(&compiled_page, (char *)b->next_to_read,  2,  match_page, 0 ) == 0) {
	  fwrite(b->next_to_read+match_page[0].rm_so,b->next_to_fill - (b->next_to_read+match_page[0].rm_so), 1, stdout);
	  b->next_to_read = b->end;
	  b->bytes_avail = 0;
	  b->next_to_fill = b->buffer; /* empty */
	  bfile.strm.next_out = (char *)b->next_to_fill;
	  bfile.strm.avail_out = b->end - b->next_to_fill;
	  firstpage = 0;
	}
	else {
	  /* could have the first part of the page tag... so copy up enough bytes to cover that case */
	  if (b->bytes_avail> 7) {
	    /* write everything that didn't match, but leave 7 bytes, to stdout */
	    fwrite(b->next_to_read,b->bytes_avail - 7,1,stdout);
	    move_bytes_to_buffer_start(b, b->next_to_read + b->bytes_avail - 7, 7);
	    bfile.strm.next_out = (char *)b->next_to_fill;
	    bfile.strm.avail_out = b->end - b->next_to_fill;
	  }
	  else {
	    if (buffer_is_empty(b)) {
	      bfile.strm.next_out = (char *)b->buffer;
	      bfile.strm.avail_out = bfile.bufout_size;
	      b->next_to_fill = b->buffer; /* empty */
	    }
	    else {
	      /* there were only 7 or less bytes so just save em don't write em to stdout */
	      move_bytes_to_buffer_start(b, b->next_to_read, b->bytes_avail);
	      bfile.strm.next_out = (char *)b->next_to_fill;
	      bfile.strm.avail_out = b->end - b->next_to_fill;
	    }
	  }
	}
      }
      else {
	if (b->bytes_avail) {
	  fwrite(b->next_to_read,b->bytes_avail,1,stdout);
	  b->next_to_read = b->end;
	  b->bytes_avail = 0;
	  b->next_to_fill = b->buffer; /* empty */
	  bfile.strm.next_out = (char *)b->next_to_fill;
	  bfile.strm.avail_out = b->end - b->next_to_fill;
	}
      }
    }
  }
  if (b->bytes_avail) {
    fwrite(b->next_to_read,b->bytes_avail,1,stdout);
    b->next_to_read = b->end;
    b->bytes_avail = 0;
    b->next_to_fill = b->buffer; /* empty */
    bfile.strm.next_out = (char *)b->next_to_fill;
    bfile.strm.avail_out = b->end - b->next_to_fill;
  }  
  return(0);
}

int main(int argc, char **argv) {
  int fin, res;
  off_t position;

  int optc;
  int optindex=0;

  struct option optvalues[] = {
    {"help", 0, 0, 'h'},
    {"version", 0, 0, 'v'},
    {NULL, 0, NULL, 0}
  };

  if (argc < 2 || argc > 3) {
    usage("Missing or bad options/arguments");
    exit(-1);
  }

  while (1) {
    optc=getopt_long_only(argc,argv,"hv", optvalues, &optindex);
    if (optc=='h')
      usage(NULL);
    else if (optc=='v')
      show_version(VERSION);
    else if (optc==-1) break;
    else usage("Unknown option or other error\n");
  }

  if (optind >= argc) {
    usage("Missing filename argument.");
  }

  fin = open (argv[optind], O_RDONLY);
  if (fin < 0) {
    fprintf(stderr,"failed to open file %s for read\n", argv[optind]);
    exit(-1);
  }
  optind++;
  if (optind >= argc) {
    usage("Missing offset argument.");
  }
  position = atoll(argv[optind]);
  if (position <(off_t)0) {
    fprintf(stderr,"please specify an offset >= 0.\n");
    fprintf(stderr,"usage: %s infile offset\n", argv[0]);
    exit(-1);
  }
  /* input file, starting position in file, length of buffer for reading */
  res = dump_mw_header(fin);

  res = dump_from_first_page_id_after_offset(fin, position);
  exit(res);
}
